pkg/service/server.go

```diff
@@ -0,0 +1,141 @@
+package service
+
+import (
+	"context"
+	"fmt"
+	"time"
+
+	"capnproto.org/go/capnp/v3"
+	"github.com/wetware/ww/internal/api/channel"
+	api "github.com/wetware/ww/internal/api/service"
+	"github.com/wetware/ww/pkg/pubsub"
+)
+
+type Server struct{}
+
+func (s Server) Registry() Registry {
+	sds := RegistryServer{}
+	return Registry(sds.Client())
+}
+
+type RegistryServer struct{}
+
+func (s *RegistryServer) Client() capnp.Client {
+	return capnp.Client(api.Registry_ServerToClient(s))
+}
+
+func (s *RegistryServer) Provide(ctx context.Context, call api.Registry_provide) error {
+	loc, err := call.Args().Location()
+	if err != nil {
+		return fmt.Errorf("failed to read location: %w", err)
+	}
+
+	response, err := encodeResponse(loc)
+	if err != nil {
+		return err
+	}
+
+	topic := pubsub.Topic(call.Args().Topic())
+
+	// subscribe to topic
+	sub, release := topic.Subscribe(ctx)
+	defer release()
+
+	call.Go()
+	for b := sub.Next(); b != nil; b = sub.Next() {
+		msg, err := decodeMessage(b)
+		if err != nil {
+			return err
+		}
+
+		if msg.Which() == api.Message_Which_request {
+			if err := topic.Publish(ctx, response); err != nil {
+				return err
+			}
+		}
+	}
+
+	return sub.Err()
+}
+
+func (s *RegistryServer) FindProviders(ctx context.Context, call api.Registry_findProviders) error {
+	request, err := encodeRequest(call)
+	if err != nil {
+		return err
+	}
+
+	topic := pubsub.Topic(call.Args().Topic())
+
+	sub, release := topic.Subscribe(ctx)
+	defer release()
+
+	time.Sleep(time.Second)
```

Comment:
What is this sleep for?

Reply:
@aratz-lasa If this is a workaround for the promised-capability issue, I would suggest doing `err := capnp.Client(topic).Resolve(ctx)` to wait until the promise has actually resolved.

(@mikelsr nice catch.)